llvm/lib/Target/NVPTX/NVPTXInstrInfo.td

```diff
@@ -2259,27 +2261,69 @@ def : Pat<(setueq Int1Regs:$a, Int1Regs:$b),
           (NOT1 (XORb1rr Int1Regs:$a, Int1Regs:$b))>;
 
 // comparisons of i8 extracted with BFE as i32
-def: Pat<(setgt (sext_inreg (trunc Int32Regs:$a), i8), (sext_inreg (trunc Int32Regs:$b), i8)),
-         (SETP_s32rr Int32Regs:$a, Int32Regs:$b, CmpGT)>;
```

Comment:
This was the root cause of the problem. We should've used `bfe($a/$b, 8, 8)`.
The tests worked because $a/$h happened to be the result of `bfe()` we've lowered for `extractement()`, so PTX output looked sane.